src/DiffRules.jl

```diff
@@ -2,6 +2,8 @@ __precompile__()
 
 module DiffRules
 
+using IrrationalConstants
```

Comment:
Correct!

What's the argument here against something like `:( inv($x) / oftype(float($x), log(10)) )`, and `:($x / 2 / pi)`?

Reply:
I would again argue, using irrationals is cleaner, simpler, and avoids unnecessary loss of precision if desired. One can avoid heuristics such as `float($x)` and instead computes exactly the desired result in the most natural and least invasive types. Moreover, e.g., `log(10)` will always be of type `Float64` and hence the reduced precision can't be recovered when working with other more accurate number types.

Reply:
IrrationalConstants is also already an indirect dependency of DiffRules, both through SpecialFunctions and LogExpFunctions, so adding it as an dependency does not affect the tree of dependencies.

Reply:
Yes I agree it looks cleaner. I worry, as before, about going from loading nothing to loading packages... which may decide they want to load a few other small packages... 

One thought is this: If SpecialFunctions loads this, then calling things as `SpecialFunctions.IrrationalConstants.twopi` is guaranteed to work, right?

Reply:
Since it's also a dependency of LogExpFunctions, the choice is a bit arbitrary... And in any case it would require that `SpecialFunctions` is available in the module where the rule is evaled - which might not be the case, e.g., if you are only interested in defining rules for functions from Base (it's easy to filter them, even more so with the new keyword argument for DiffRules).

I think this example highlights that in fact it might be better to interpolate all such references in the expressions - to irrationals but also to functions from e.g. SpecialFunctions (and hence to load SpecialFunctions). It's less likely that evaluating rules leads to undefined references and hence seems much safer in general. I'm still not convinced that avoiding `using IrrationalConstants` or `using SpecialFunctions` is really desirable - if these packages are broken most downstream packages that use DiffRules will be broken anyway (and these packages are so central that it would affect large parts of the Julia ecosystem) and hence I don't think one has to be concerned with possible upstream issues when loading the packages. Whereas I think DiffRules should be concerned with issues caused by non-interpolated local references in the rules.

Reply:
Any updates on this?

Reply:
My opinion is that we should change old packages like this as little as possible, at least not without serious gain. E.g. "interpolate all such references" might have been a better design, but isn't the one chosen, and changing all of them sounds like a way to create unexpected headaches, and gain nothing.

I also don't think it should go from loading zero dependencies to loading a bunch, without serious gain. IIRC there are exactly 4 cases where writing `SpecialFunctions.IrrationalConstants.sqrtπ` would not work, and they matter only if you happen to want to AD a few things like `log10(::BigFloat)` efficiently. That seems pretty obscure.

Reply:
> and changing all of them sounds like a way to create unexpected headaches, and gain nothing.

I would argue that the gain is not nothing. I have code that runs 2X faster by using this branch vs. previous release when using Zygote (it's a pretty high-dim problem). And possibly even more significant: last time I checked the lack of proper conversion completely breaks broadcasting over `ForwardDiff.Dual` on GPU, which also means that Zygote breaks when you try to broadcast over some of these functions.

Reply:
No, sorry, I don't mean that types don't matter. I mean that changing from an Expr containing a Symbol, to one containing a Function, doesn't matter, as the `@eval`'d code ends up the same.

Reply:
Ah, gotcha :+1:

> I don't see any such changes in this PR right now.

So you're happy with the `$logten`, etc. that I've done but don't want to start interpolating _everything_?

Reply:
And the only issue that is holding back this PR atm is whether we should depend directly or indirectly on `IrrationalConstants`?

Reply:
I'm against adding dependencies. I believe that you can have all the benefits of `sqrtπ` without explicitly loading the package, which you must do to interpolate `:($sqrtπ)` from this package's scope. 

It seems reasonable to assume that any user of the SpecialFunctions rules must have SpecialFunctions loaded, and thus I don't see a problem with accessing constants within it: I find it hard to imagine where `:( 2 * (($x * SpecialFunctions.erfcx($x)) - $invsqrtπ) )` would work but something like  `:( 2 * (($x * SpecialFunctions.erfcx($x)) - SpecialFunctions.erfcx.invsqrtπ) )` would not. 

I don't think `:( inv($logten * $x) )` is worth a dependency, compared to `:( inv($logten * log(oftype($x, 10))) )`.

Reply:
I think interpolating everything is the cleaner and safer way but out of scope for this PR. IMO, however, we should load and interpolate the irrational constants instead of relying on workarounds that are trivial to break and are inconsistent with our API that allows to filter rules for specific packages if only a subset of them is loaded.

Reply:
How is this inconsistent with the API? If you have loaded SpecialFunctions, then `@eval` will work. If not, not.

Reply:
It is inconsistent if we use SpecialFunctions.IrrationalConstants... in rules for functions that are not defined in SpecialFunctions.

I am very strongly against SpecialFunctions.IrrationalConstants... since it is a hack without IMO clear benefit but sever disadvantages: If SpecialFunctions decides to drop the dependency on IrrationalConstants, which can be done in any non-breaking release, then DiffRules and all downstream packages will be broken. And in fact, this PR will break all downstream packages and uses for SpecialFunctions < 1.7.0 since only in this release the dependency on IrrationalConstants was introduced in SpecialFunctions. Ie we would have to bump the compat without any other reason, and could still not be sure if SpecialFunctions 2.1.0 breaks the whole DiffRules ecosystem.

Reply:
> in rules for functions that are not defined in SpecialFunctions.

Yes, there are 4 such, which only matter for some strange BigFloat case.

> If SpecialFunctions decides to drop the dependency on IrrationalConstants, which can be done in any non-breaking release

I agree this trick upgrades that to part of SpecialFunctions's API, unavoidably.

Dropping SpecialFunctions 0.10 seems like no problem. Why is a lower bound at 1.7.0 going to break anything?

Reply:
> Yes, you can't do that. There are 4 such in the PR, which only matter for some strange BigFloat case, and can just be reverted.

Am a bit confused by this statement. If we remove the usage of irrational constants, won't the `log10` rule result in promotion to `Float64` from `Float32`? This doesn't seem like a "strange case".

Reply:
The alternative to a fancy log10 is `log(oftype(x, 10)`. Which compiles to the exact same code. Except, I think, for `x::BigFloat` which is mutable. (But slow in any case, and probably has not occurred in the wild.)

Reply:
Ah gotcha :+1: 

I'm still confused as to why it's preferable to depend on IrrationalConstants implicitly vs. explicitly, given that explicit dependence allows us to control our dependence on IrrationalConstants directly (and thus avoiding issues/annoyances as outlined by @devmotion above) rather than relying on SpecialFunctions not doing something surprising.

> I agree this trick upgrades that to part of SpecialFunctions's API, unavoidably.

But isn't it avoidable by just depending on IrrationalConstants directly?

Reply:
The point of the trick is to avoid *loading* any packages. That's one meaning of dependency. And it's the meaning under which Static.jl managed to break about a thousand packages which didn't know they depended on it, last year. That's the sort of britleness I'd like to avoid having more of. I suspect this is a losing battle, and soon everything will depend on everything else, but I try.

This package still *constrains* SpecialFunctions via compat, which is read by Pkg without loading any code. That's another meaning of dependency, and would have avoided the catastrophe. This package could similarly constrain IrrationalConstants if this seems like a good idea. Adding that would be free, since (for suitably new SpecialFunctions) Pkg already downloads the package.

Reply:
I disagree, I think making it an implicit dependency makes DiffRules very brittle because every new SpecialFunctions release may break DiffRules without us having any control over it. That seems much more brittle than potential issues with `using IrrationalConstants`. IrrationalConstants is such a central package in the ecosystem that I don't think DiffRules can avoid downstream errors by not loading it if e.g. SpecialFunctions, LogExpFunctions, StatsFuns etc all depend on it and load it.

Reply:
Oooh I didn't realize DiffRules no longer loads SpecialFunctions.. The discussion makes more sense now :+1: 

Well, though I'm still uncertain which approach is preferred, I'm quite keen on getting this through, and so I'm fine with what you're suggesting :+1: 

Btw, is there a particular reason why we shouldn't just move the rules for SpecialFunctions into, well, SpecialFunctions itself?

Reply:
I think @mcabbott 's point is that diffrules for SpecialFunctions will only be evaluated in a scope where SpecialFunctions is available, and so if indeed a SpecialFunctions release breaks the diffrules for methods from SpecialFunctions, doing `using DiffRules` will still work if you're not calling `@eval` on any of these particular rules.

Reply:
> I think @mcabbott 's point is that diffrules for SpecialFunctions will only be evaluated in a scope where SpecialFunctions is available, and so if indeed a SpecialFunctions release breaks the diffrules for methods from SpecialFunctions, doing `using DiffRules` will still work if you're not calling `@eval` on any of these particular rules. 
> 
> I.e. SpecialFunctions will only break rules for SpecialFunctions.

Sure, but how is this not the responsibility of DiffRules which defines these brittle rules in the first place? In such a scenario the only proper fix is to change the definition of the rule in DiffRules. I still don't understand why we should risk breaking the whole AD ecosystem instead of loading a lightweight dependency that is loaded already by AD packages but also other core packages, directly or indirectly.

Reply:
> Btw, is there a particular reason why we shouldn't just move the rules for SpecialFunctions into, well, SpecialFunctions itself?

I think this might be a limitation of the design of ForwardDiff (and perhaps others). The `eval` code on loading, and never look again, so rules which aren't in the list at first won't be loaded. Certainly that's true for independent packages loaded later (ChainRules is invented partly to solve this problem) but there's a chance that rules defined in SpecialFunctions would make it, so long as `using SpecialFuncitons` is above the `eval` loop in ForwardDiff? Haven't tried. Might break other things.

> SpecialFunctions will only break rules for SpecialFunctions

Yes. Downstream users of ForwardDiff who only want Base functions need never know if things are (temporarily) broken, because defining a function which calls a nonexistent symbol isn't an error (until you call it).

I guess the chance of IrrationalFunctions causing a Static.jl-style apocaplyse are low, since it's much smaller, simpler. Although I thought this about EllipsisNotation, too, which had been the same 5 lines (zero dependencies) for years and years, and then somehow acquired a dependence which had dependants and suddenly it's many thousands of lines and Base.Experimental macros and all sorts of weird highly optimised stuff. 

Anyway, I have probably said too much, and should shut up. Maybe I unsubscribe. If this PR does want to go from zero to one full-time dependencies, I do think that's something another maintainer ought to sign off on. Sorry if this is more of a rat's nest than you wanted to get into, Tor, thanks for the work.

Reply:
> I think this might be a limitation of the design of ForwardDiff (and perhaps others). The eval code on loading, and never look again, so rules which aren't in the list at first won't be loaded.

Yeah I just remembered how ForwardDiff's mechanism works. If packages were to define their own rules, they'd have to also load ForwardDiff.

> Yes. Downstream users of ForwardDiff who only want Base functions need never know if things are (temporarily) broken, because defining a function which calls a nonexistent symbol isn't an error (until you call it).

But are there many packages that use DiffRules in this way? Maybe SymbolicUtils.jl? But most dependance on DiffRules comes through ForwardDiff, which already loads SpecialFunctions, etc.

> I guess the chance of IrrationalFunctions causing a Static.jl-style apocaplyse are low, since it's much smaller, simpler. Although I thought this about EllipsisNotation, too, which had been the same 5 lines (zero dependencies) for years and years, and then somehow acquired a dependence which had dependants and suddenly it's many thousands of lines and Base.Experimental macros and all sorts of weird highly optimised stuff.

These experiences with Static.jl and EllipsisNotation.jl you're talking about, was this prior to compat-entries in Package.toml? 

> If this PR does want to go from zero to one full-time dependencies, I do think that's something another maintainer ought to sign off on. 

As I've said, I'm happy to go with the zero dependency approach now that I've realized this is the path DiffRules has already gone down. Whether or not this should be the case, should maybe be left for as a separate issue/PR?

> Sorry if this is more of a rat's nest than you wanted to get into, Tor, thanks for the work.

No worries! I didn't have expectations of this being just a "quick patch" exactly due to how much code depends on DiffRules. And thank you for explaining everything in great detail and being so patient with me :+1: 

_So_. How about we take the approach of _not_ loading IrrationalConstants _in this PR_, and defer introduction of loading IrrationalConstants to a separate issue/PR? I think the fixes this PR has is of sufficient value that it's worth just getting it in "ASAP".

Reply:
Okay, so one problem: SpecialFunctions.jl only imports the variables it explicitly needs https://github.com/JuliaMath/SpecialFunctions.jl/blob/414385a7097e0f376b1273a3e0deac9d30cf0f9e/src/SpecialFunctions.jl#L3-L14

This in turn means that we can't just do `SpecialFunctions.IrrationalConstants.sqrtπ`, but instead have to do `SpecialFunctions.sqrtπ`. Now this is fine _for now_, but will break if SpecalFunctions decides to change this and/or we need some constants that aren't explicitly used in SpecialFunctions.

Reply:
> As I've said, I'm happy to go with the zero dependency approach now that I've realized this is the path DiffRules has already gone down. Whether or not this should be the case, should maybe be left for as a separate issue/PR?

DiffRules also loaded e.g. LogExpFunctions in some releases (to figure out if some functions are available), this was changed only recently. I think there's a significant difference between both packages even though both are used by SpecialFunctions - IrrationalConstants is much more central and lightweight and only adds constants that are not in base (by committing type piracy technically but this seems OK - and definitely better than defining these constants in multiple packages since then you end up with many method redefinition warnings). Hence as an extension of base it's less surprising that it is useful even for derivatives of functions from base.

Reply:
>  I think there's a significant difference between both packages even though both are used by SpecialFunctions - IrrationalConstants is much more central and lightweight and only adds constants that are not in base (by committing type piracy technically but this seems OK - and definitely better than defining these constants in multiple packages since then you end up with many method redefinition warnings). Hence as an extension of base it's less surprising that it is useful even for derivatives of functions from base.

Just for the record: I'm also in favour of loading IrrationalConstants.jl, but my opinion isn't so important here given that I'm not one of the maintainers of DiffRules :shrug: At this point, I just want to get this PR merged. _But_, as I pointed out above, the "naive" way of just using whatever is present in SpecialFunctions seems _very_ brittle.

Reply:
I guess @mcabbott and I both stated and explained our preferences, so maybe it would be good to include some other contributor. @andreasnoack, it seems you commented on an earlier draft of this PR - what's your opinion? Should we depend on, load IrrationalConstants, and interpolate constants or refer to (unexported) constants in SpecialFunctions without depending and loading IrrationalConstants (and use them only for functions from SpecialFunctions)?

Reply:
Just checking up on this again. Any conclusion to this? Are we still waiting for another maintainer to respond or has it already been decided that it's a "nah"?

Reply:
Since nobody has replied, let's just do it in the controlled future-proof way and load and interpolate the constants. Since we already depend on it indirectly, it shouldn't cause any problems (BTW in the mean time also CR added it as a dependency). And it's easy to change if we want to in the future and there's no risk of SpecialFunctions breaking anything accidentally.

Reply:
That is, as it's currently implemented? Sooo we merge?